File: assignment 7/python/p3.py
#Write a program in python to check whether number of opening'(' and closing')' parentheses are equal or not in an expression using stack.
class Stack:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        return None

def check_parentheses(expression):
    stack = Stack()
    for char in expression:
        if char == '(':
            stack.push(char)
        elif char == ')':
            if stack.is_empty():
                return False
            stack.pop()
    return stack.is_empty()

# Example usage
if __name__ == "__main__":
    expr = input("Enter an expression: ")
    if check_parentheses(expr):
        print("Number of opening and closing parentheses are equal.")
    else:
        print("Unmatched parentheses found.")

File: assignment 7/c programming/p3.c
//Write a program in C to check whether number of opening'(' and closing')' parentheses are equal or not in an expression using stack.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 100

char stack[MAX];
int top = -1;

void push(char c) {
    if (top < MAX - 1) {
        stack[++top] = c;
    }
}

char pop() {
    if (top >= 0) {
        return stack[top--];
    }
    return '\0';
}

int is_empty() {
    return top == -1;
}

int main() {
    char expression[MAX];
    printf("Enter an expression: ");
    fgets(expression, MAX, stdin);
    expression[strcspn(expression, "\n")] = 0;  // Remove newline character

    for (int i = 0; i < strlen(expression); i++) {
        if (expression[i] == '(') {
            push('(');
        } else if (expression[i] == ')') {
            if (is_empty()) {
                printf("Unmatched closing parenthesis found.\n");
                return 0;
            }
            pop();
        }
    }

    if (is_empty()) {
        printf("Number of opening and closing parentheses are equal.\n");
    } else {
        printf("Unmatched opening parenthesis found.\n");
    }

    return 0;
}
